### PR TITLE
Add VmomiSupport.VmomiJSONEncoder for encoding pyVmomi objects as JSON

### LINKED ISSUE #21: Serialize to JSON?
Hi,

I was wondering whether it is possible to serialize an object properties to JSON?

For example taking the following code snippet, how could I dump the `vm.summary` properties to JSON?

```
>>> from pyVim.connect import SmartConnect, Disconnect
>>> si = SmartConnect(host='vc01', user='root', pwd='password')
>>> datacenter = content.rootFolder.childEntity[0]
>>> vmFolder = datacenter.vmFolder
>>> vmList = vmFolder.childEntity
>>> vm = vmList[1].childEntity[1]
>>> vm
'vim.VirtualMachine:vm-182'
>>> vm.summary
```

Having a JSON serializer would really make things simple and easier to integrate with other systems. 

Is this possible using pyVmomi?

Thanks and regards,
Marin

### PR DESCRIPTION
**Pull request by jeking3**
This fixes #21 
This closes #669 

This allows for deep JSON encoding of pyVmomi Managed Objects by providing a JSONEncoder.  Tested with Datacenter, Datastore, HostSystem, Network, VirtualMachine objects against a VCSA version 6.7.  A sample python script is provided allowing any object to be dumped by type and MOID.

#### Example: Serialize a VirtualMachine
```
    data = json.dumps(vm, cls=VmomiJSONEncoder)
```

#### Example: Serialize a list of VirtualMachines and their snapshots
```
    data = json.dumps([vm1, vm2], cls=VmomiJSONEncoder,
                      explode=[templateOf('VirtualMachine'),
                               templateOf('VirtualMachineSnapshot')])
```

##### Command Line
Assuming you have already run tox:
```
~/pyvmomi$ PYTHONPATH=.:.tox/py27/lib/python2.7/site-packages python sample/dumpjson.py \
    --host vcenter --user my_user --pass my_pass --type Network --id network-16
```
##### Code
Or perhaps you just want to see the code:
```
   obj = VmomiSupport.templateOf('Network')('network-16', si._stub)
   print(json.dumps(obj, cls=VmomiSupport.VmomiJSONEncoder,
                    sort_keys=True, indent=2))
```
##### Result
Note this is perhaps a poor example, since the encoder handles deep structures completely, but the other examples are rather huge.
```
{
  "_vimid": "network-16",
  "_vimref": "vim.Network:network-16",
  "_vimtype": "vim.Network",
  "alarmActionsEnabled": true,
  "availableField": [],
  "configIssue": [],
  "configStatus": "green",
  "customValue": [],
  "declaredAlarmState": [],
  "disabledMethod": [],
  "effectiveRole": [
    -1
  ],
  "host": [
    "vim.HostSystem:host-14",
    "vim.HostSystem:host-25"
  ],
  "name": "VM Network",
  "overallStatus": "green",
  "parent": "vim.Folder:group-n6",
  "permission": [],
  "recentTask": [],
  "summary": {
    "accessible": true,
    "dynamicProperty": [],
    "dynamicType": null,
    "ipPoolId": null,
    "ipPoolName": "",
    "name": "VM Network",
    "network": "vim.Network:network-16"
  },
  "tag": [],
  "triggeredAlarmState": [],
  "value": [],
  "vm": [
    "vim.VirtualMachine:vm-17",
    "vim.VirtualMachine:vm-18",
    "vim.VirtualMachine:vm-19",
    "vim.VirtualMachine:vm-20",
    "vim.VirtualMachine:vm-21",
    "vim.VirtualMachine:vm-22",
    "vim.VirtualMachine:vm-27",
    "vim.VirtualMachine:vm-28",
    "vim.VirtualMachine:vm-29",
    "vim.VirtualMachine:vm-32",
    "vim.VirtualMachine:vm-36",
    "vim.VirtualMachine:vm-42",
    "vim.VirtualMachine:vm-69",
    "vim.VirtualMachine:vm-86",
    "vim.VirtualMachine:vm-87",
    "vim.VirtualMachine:vm-111",
    "vim.VirtualMachine:vm-112"
  ]
}
```

### CODE REVIEW

**Comment by jeking3** [comment]
I didn't change anything relative to SSL, so it's odd that two SSL tests are failing.  @tianhao64 this also happens locally for me.  I'll take another look as to why.

**Comment by jeking3** [comment]
@pgbidkar when will the next release be?

**Comment by tianhao64** [comment]
@jeking3 Looks like the vcrpy latest release is not working well with existing SSL tests. If you pull the latest code or update vcrpy==1.12.0 in test-requirements.txt, the tests should pass.

**Comment by jeking3** [comment]
@tianhao64 I did that (actually set it to `vcrpy>2.0.0` which seems to work just fine, since 1.13.0 is out and worked for me).

I have pull requests in other projects that depends on this change getting merged and released.  I think it provides significant value.  Let me know if anything is blocking this from getting in.

File: pyVmomi/VmomiSupport.py [hunk 1]
```diff
@@ -273,6 +274,106 @@ def __call__(self, **kwargs):
       else:
          raise AttributeError("'%s' does not exist" % self.name)
 
+
+class VmomiJSONEncoder(json.JSONEncoder):
+    """
```
Comment (pgbidkar) [inline pyVmomi/VmomiSupport.py]: Indentation in pyVmomi code is 3 space. Looks like you did 4 space indentation.
Please can you check and change that?

**Comment by jeking3** [comment]
@pgbidkar I do not see what you are seeing.  This works fine for me:
```
pyvmomi$ PYTHONPATH=.:.tox/py27/lib/python2.7/site-packages/ python sample/dumpjson.py -s 10.0.0.4 -u my_user -p my_pass -t VirtualMachine -i vm-22
```
There must be something different in your environment.  Are you running against a real vCenter or a simulator?

**Comment by jeking3** [comment]
@pgbidkar 3 space indents are non-standard, you should consider standardizing.  I made the changes you requested.

https://lintlyci.github.io/Flake8Rules/rules/E111.html

**Comment by pgbidkar** [comment]
Hi @jeking3 , I am running sample with Real vCenter

**Comment by jeking3** [comment]
Not sure where you got "dump_all_json.py" from; the sample included is called "dumpjson.py".  Line 345 is not the proper line for calling to default, like 352 is.  Do you have your virtual environment set up with the latest copy of this PR?

**Comment by jeking3** [comment]
Any thoughts?

**Comment by pgbidkar** [comment]
Merging the change. Thanks

**Comment by jeking3** [comment]
Thanks.  Now the only thing blocking upstream changes is a pip release.

**Pull request merged**